### DISCUSSION TITLE
List typeof DocumentLibrary: get groups and role (assignments and definitions)

### DISCUSSION THREAD

**Discussion by gonzaloleon** [pnp/pnpcore#931] [Q&A] [answered]
Hey there!
I'm trying to replicate the code here: https://www.sharepointdiary.com/2018/09/sharepoint-online-get-user-permission-report-using-powershell.html but with .NET, and I'm stuck on getting DocumentLibraries groups and roles when they hasUniqueRoleAssignments.

We have some Sharepoint sites, I'm able to loop through all sites to get Roles and permissions (also I can see the group with Full Control permissions on a particular Document Library with "Limited Access" role in the site permissions, I think that's not bad).

Use case:
- Within a Site, create a Document Library
- Into that Document Library, click in the gear (upper right corner), "Library Settings" and then "Permissions for this document library"
- Grant Permissions and add a group with "full control"

Code I'm using:

`foreach (var list in Web.Lists.Where(r => r.TemplateType == PnP.Core.Model.SharePoint.ListTemplateType.DocumentLibrary && !r.Hidden))
{
    if (list.HasUniqueRoleAssignments)
    {
        // TODO: get groups -> ???
    } else {
       // Same groups and roles as parent.
    }
}
`

I hope you can help me to solve this.
Maybe I misunderstood something.
Thanks!

**Comment by jansenbe** [comment, marked as answer]
@gonzaloleon : Below code snippet gets the document libraries and checks for unique permissions. Your code missed the `QueryProperties` which is needed as by default the `HasUniqueRoleAssignment` property is not returned by SharePoint.

```csharp
context.Web.Load(p => p.RoleAssignments.QueryProperties(p => p.RoleDefinitions, p => p.PrincipalId));

foreach (var list in context.Web.Lists.QueryProperties(p => p.TemplateType, p => p.HasUniqueRoleAssignments, p => p.Title, 
                                                       p => p.RoleAssignments.QueryProperties(p => p.RoleDefinitions, p => p.PrincipalId))
                                      .Where(r => r.TemplateType == ListTemplateType.DocumentLibrary && !r.Hidden))
{

    if (list.HasUniqueRoleAssignments)
    {
        // Get the custom role assignments for this list
        foreach (var roleAssignment in list.RoleAssignments.AsRequested())
        {
            // do something with the role assignment (principal granted role definitions)
        }
    }
    else
    {
        // List uses the role assignments of the web
        foreach(var roleAssignment in context.Web.RoleAssignments.AsRequested())
        {
            // do something with the role assignment (principal granted role definitions)
        }
    }
}
```

**Reply by gonzaloleon** [reply]
> @gonzaloleon : Below code snippet gets the document libraries and checks for unique permissions. Your code missed the `QueryProperties` which is needed as by default the `HasUniqueRoleAssignment` property is not returned by SharePoint.
> 
> ```cs
> context.Web.Load(p => p.RoleAssignments.QueryProperties(p => p.RoleDefinitions, p => p.PrincipalId));
> 
> foreach (var list in context.Web.Lists.QueryProperties(p => p.TemplateType, p => p.HasUniqueRoleAssignments, p => p.Title, 
>                                                        p => p.RoleAssignments.QueryProperties(p => p.RoleDefinitions, p => p.PrincipalId))
>                                       .Where(r => r.TemplateType == ListTemplateType.DocumentLibrary && !r.Hidden))
> {
> 
>     if (list.HasUniqueRoleAssignments)
>     {
>         // Get the custom role assignments for this list
>         foreach (var roleAssignment in list.RoleAssignments.AsRequested())
>         {
>             // do something with the role assignment (principal granted role definitions)
>         }
>     }
>     else
>     {
>         // List uses the role assignments of the web
>         foreach(var roleAssignment in context.Web.RoleAssignments.AsRequested())
>         {
>             // do something with the role assignment (principal granted role definitions)
>         }
>     }
> }
> ```

@jansenbe thank you so much for the answer!
One more thing, please:

How could I get the groups and privilege levels here?

```
if (list.HasUniqueRoleAssignments)
{
        foreach (var roleAssignment in list.RoleAssignments.AsRequested())
        {
            // do something with the role assignment (principal granted role definitions)
        }
}
```

Thanks!

**Reply by jansenbe** [reply]
@gonzaloleon : here you go :-)

```csharp
// Load the site groups and role assigments in one go
context.Web.Load(p => p.SiteGroups.QueryProperties(p => p.Users, p => p.Title),
                 p => p.SiteUsers.QueryProperties(p => p.Id, p => p.Title),
                 p => p.RoleAssignments.QueryProperties(p => p.RoleDefinitions, p => p.PrincipalId));

foreach (var list in context.Web.Lists.QueryProperties(p => p.TemplateType, p => p.HasUniqueRoleAssignments, p => p.Title,
                                                       p => p.RoleAssignments.QueryProperties(p => p.RoleDefinitions, p => p.PrincipalId))
                                      .Where(r => r.TemplateType == ListTemplateType.DocumentLibrary && !r.Hidden))
{

    if (list.HasUniqueRoleAssignments)
    {
        // Get the custom role assignments for this list
        foreach (var roleAssignment in list.RoleAssignments.AsRequested())
        {
            // do something with the role assignment (principal granted role definitions)
        }
    }
    else
    {
        // List uses the role assignments of the web
        foreach (var roleAssignment in context.Web.RoleAssignments.AsRequested())
        {
            // do something with the role assignment (principal granted role definitions)
            var group = context.Web.SiteGroups.AsRequested().FirstOrDefault(g => g.Id == roleAssignment.PrincipalId);
            if (group != null)
            {
                // roleAssignment.RoleDefinitions contains the role definitions granted to the group

                foreach (var user in group.Users.AsRequested())
                {
                    // list the users in the group
                }
            }
            else
            {
                var user = context.Web.SiteUsers.AsRequested().FirstOrDefault(u => u.Id == roleAssignment.PrincipalId);
                if (user != null)
                {
                    // roleAssignment.RoleDefinitions contains the role definitions granted to the user
                }
            }
        }
    }
}
```

**Reply by gonzaloleon** [reply]
> @gonzaloleon : here you go :-)
> 
> ```cs
> // Load the site groups and role assigments in one go
> context.Web.Load(p => p.SiteGroups.QueryProperties(p => p.Users, p => p.Title),
>                  p => p.SiteUsers.QueryProperties(p => p.Id, p => p.Title),
>                  p => p.RoleAssignments.QueryProperties(p => p.RoleDefinitions, p => p.PrincipalId));
> 
> foreach (var list in context.Web.Lists.QueryProperties(p => p.TemplateType, p => p.HasUniqueRoleAssignments, p => p.Title,
>                                                        p => p.RoleAssignments.QueryProperties(p => p.RoleDefinitions, p => p.PrincipalId))
>                                       .Where(r => r.TemplateType == ListTemplateType.DocumentLibrary && !r.Hidden))
> {
> 
>     if (list.HasUniqueRoleAssignments)
>     {
>         // Get the custom role assignments for this list
>         foreach (var roleAssignment in list.RoleAssignments.AsRequested())
>         {
>             // do something with the role assignment (principal granted role definitions)
>         }
>     }
>     else
>     {
>         // List uses the role assignments of the web
>         foreach (var roleAssignment in context.Web.RoleAssignments.AsRequested())
>         {
>             // do something with the role assignment (principal granted role definitions)
>             var group = context.Web.SiteGroups.AsRequested().FirstOrDefault(g => g.Id == roleAssignment.PrincipalId);
>             if (group != null)
>             {
>                 // roleAssignment.RoleDefinitions contains the role definitions granted to the group
> 
>                 foreach (var user in group.Users.AsRequested())
>                 {
>                     // list the users in the group
>                 }
>             }
>             else
>             {
>                 var user = context.Web.SiteUsers.AsRequested().FirstOrDefault(u => u.Id == roleAssignment.PrincipalId);
>                 if (user != null)
>                 {
>                     // roleAssignment.RoleDefinitions contains the role definitions granted to the user
>                 }
>             }
>         }
>     }
> }
> ```

thanks, I already solved that for the same users & groups as the site, but what about when the list HasUniqueRoleAssignments?
Could we get users & groups (mainly groups)?

**Reply by gonzaloleon** [reply]
I'm stuck when the list HasUniqueRoleAssignments, I can't get groups with role definitions and assignments there.

**Reply by jansenbe** [reply]
@gonzaloleon : those groups live at web level, you can use the PrincipalID of the role assignment of a list to find the respective group

**Reply by gonzaloleon** [reply]
> @gonzaloleon : those groups live at web level, you can use the PrincipalID of the role assignment of a list to find the respective group

so the PrincipalID refers to a group in context.Web.SiteGroups?

**Reply by jansenbe** [reply]
@gonzaloleon : yes, to a group (SiteGroups) or a user (SiteUsers)

**Reply by gonzaloleon** [reply]
> @gonzaloleon : yes, to a group (SiteGroups) or a user (SiteUsers)

Thank you so much for your help!
